utils/tfw_logger.cc

```diff
@@ -0,0 +1,346 @@
+
+	do {
+		if (size < (int)sizeof(TfwBinLogEvent))
+			return;
+
+		event = (TfwBinLogEvent *)p;
+
+		switch (event->type) {
+		case TFW_MMAP_LOG_TYPE_ACCESS:
+			r = read_access_log_event(p, size, clickhouse);
+			if (r < 0)
+				return;
+			size -= r;
+			p += r;
+			break;
+		default:
+			throw Except("Unsupported log type: {}",
+				     (unsigned int)event->type);
+			return;
+		}
+	} while (size > 0);
+
+	clickhouse->commit();
+}
+
+void
+run_thread(int ncpu, int fd, std::string host, std::promise<void> promise)
+try {
+	cpu_set_t cpuset;
+	pthread_t current_thread = pthread_self();
+
+	TfwClickhouse clickhouse(host, TABLE_NAME, make_block);
+
+	TfwMmapBufferReader mbr(ncpu, fd, &clickhouse, callback);
+
+	CPU_ZERO(&cpuset);
+	CPU_SET(mbr.get_cpu_id(), &cpuset);
+
+	assert(pthread_setaffinity_np(current_thread,
+				      sizeof(cpu_set_t), &cpuset) == 0);
+
+	mbr.run(&stop_flag);
+
+	promise.set_value();
+}
+catch (...) {
+	promise.set_exception(std::current_exception());
+}
+
+void
+sig_handler([[maybe_unused]] int  sig_num) noexcept
+{
+	stop_flag = true;
+}
+
+void
+set_sig_handlers()
+{
+	struct sigaction sa;
+
+	sigemptyset(&sa.sa_mask);
+	sigaddset(&sa.sa_mask, SIGHUP);
+	sigaddset(&sa.sa_mask, SIGINT);
+	sigaddset(&sa.sa_mask, SIGTERM);
+
+	sa.sa_handler = sig_handler;
+	sa.sa_flags = SA_RESTART;
+
+	sigaction(SIGHUP, &sa, NULL);
+	sigaction(SIGINT, &sa, NULL);
+	sigaction(SIGTERM, &sa, NULL);
+}
+
+int
+main(int argc, char* argv[])
+try {
+	std::vector<std::thread> thrs;
+	std::vector<std::future<void>> futures;
+	unsigned int i;
+	long int cpu_cnt;
+	int fd;
+
+	po::options_description desc{"Usage: tfw_logger [options] <host>"};
+	desc.add_options()
+		("help,h", "show this message and exit")
+		("ncpu,n", po::value<unsigned int>(),
+			   "manually specifying the number of CPUs")
+		("host,H", po::value<std::string>(),
+			   "clickserver host address (required)")
+		;
+	po::positional_options_description pos_desc;
+	pos_desc.add("host", -1);
+	po::variables_map vm;
+	po::store(po::command_line_parser(argc, argv)
+		  .options(desc)
+		  .positional(pos_desc)
+		  .run(),
+		  vm);
+	po::notify(vm);
+	if (vm.count("help") || !vm.count("host")) {
+		std::cout << desc << std::endl;
+		return 0;
+	}
+
+	if (vm.count("ncpu")) {
+		cpu_cnt = vm["ncpu"].as<unsigned int>();
+	} else {
+		cpu_cnt = sysconf(_SC_NPROCESSORS_ONLN);
+		if (cpu_cnt < 0)
+			throw Except("Can't get CPU number");
+	}
+
+	set_sig_handlers();
+
+	while ((fd = open(FILE_PATH, O_RDWR)) == -1) {
+		if (stop_flag.load(std::memory_order_acquire))
+			return 0;
+		if (errno != ENOENT)
+			throw Except("Can't open device");
+		sleep(WAIT_FOR_FILE);
+	}
+
+	for (i = 0; i < cpu_cnt; ++i) {
+		std::promise<void> promise;
+		futures.push_back(promise.get_future());
+		thrs.push_back(std::thread(run_thread, i, fd,
+					   std::move(vm["host"].as<std::string>()),
+					   std::move(promise)));
+	}
+
+	for (i = 0; i < cpu_cnt; ++i)
+		thrs[i].join();
+
+	close(fd);
```

Comment:
It seems that we should close fd before `return 0` and in case when we catch exceptions. Because push_back can throw